samples/nrf9160/nrf_cloud_rest_agps/src/main.c

```diff
@@ -0,0 +1,604 @@
+		LOG_DBG("LTE_LC_EVT_NW_REG_STATUS: %d", evt->nw_reg_status);
+		if ((evt->nw_reg_status != LTE_LC_NW_REG_REGISTERED_HOME) &&
+		     (evt->nw_reg_status != LTE_LC_NW_REG_REGISTERED_ROAMING)) {
+			break;
+		}
+
+		LOG_INF("Network registration status: %s",
+			evt->nw_reg_status == LTE_LC_NW_REG_REGISTERED_HOME ?
+			"Connected - home network" : "Connected - roaming");
+		k_sem_give(&lte_connected);
+		break;
+	case LTE_LC_EVT_CELL_UPDATE:
+		LOG_DBG("LTE cell changed: Cell ID: %d, Tracking area: %d",
+			evt->cell.id, evt->cell.tac);
+		current_cell.id = evt->cell.id;
+		current_cell.tac = evt->cell.tac;
+		break;
+	case LTE_LC_EVT_LTE_MODE_UPDATE:
+	case LTE_LC_EVT_PSM_UPDATE:
+	case LTE_LC_EVT_EDRX_UPDATE:
+	case LTE_LC_EVT_RRC_UPDATE:
+	default:
+		break;
+	}
+}
+
+static int connect_to_network(void)
+{
+	int err;
+
+	LOG_INF("Waiting for network..");
+
+	(void)k_sem_take(&lte_connected, K_NO_WAIT);
+
+	err = lte_lc_init_and_connect_async(lte_handler);
+	if (err) {
+		LOG_ERR("Failed to init modem, err %d", err);
+	} else {
+		k_sem_take(&lte_connected, K_FOREVER);
+		LOG_INF("Connected");
+		(void)set_led(1);
+	}
+
+	return err;
+}
+
+
+void main(void)
+{
+	LOG_INF("Starting \"%s\"..", log_strdup(CONFIG_REST_SAMPLE_NAME));
+
+	char rx_buf[REST_RX_BUF_SZ];
+	int agps_sz;
+	struct jwt_data jwt = {
+#if defined(CONFIG_REST_ID_SRC_INTERNAL_UUID)
+		.subject = NULL,
+#else
+		.subject = device_id,
+#endif
+		.audience = NULL,
+		.exp_delta_s = 0,
+		.sec_tag = CONFIG_REST_JWT_SEC_TAG,
+		.key = JWT_KEY_TYPE_CLIENT_PRIV,
+		.alg = JWT_ALG_TYPE_ES256,
+		.jwt_buf = NULL
+	};
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = true,
+		.timeout_ms = 30000,
+		.rx_buf = rx_buf,
+		.rx_buf_len = sizeof(rx_buf),
+		.fragment_size = 0
+	};
+	struct lte_lc_cells_info net_info = {
+		.ncells_count = 0,
+		.neighbor_cells = NULL
+	};
+	struct nrf_cloud_rest_agps_request agps_req = {
+		.type = NRF_CLOUD_REST_AGPS_REQ_CUSTOM,
+		.agps_req = &agps_request,
+		.net_info = &net_info
+	};
+	struct nrf_cloud_rest_agps_result agps_result = {
+		.buf_sz = 0,
+		.buf = NULL
+	};
+	int err;
+
+retry:
+	err = init();
+	if (err) {
+		LOG_ERR("Initialization failed");
+		goto cleanup;
+	}
+
+	err = connect_to_network();
+	if (err) {
+		goto cleanup;
+	}
+
+	err = get_modem_info(&modem_info);
+	if (err) {
+		goto cleanup;
+	}
+	current_cell.mcc = modem_info.network.mcc.value;
+	current_cell.mnc = modem_info.network.mnc.value;
+
+	agps_print_enable(false);
+
+	err = gps_start(gps_dev, &gps_cfg);
+	if (!err) {
+		LOG_INF("GPS STARTED!");
+	}
+
+	(void)do_jitp();
+
+agps_wait:
+	LOG_INF("Waiting for AGPS request event..");
+	k_sem_take(&do_agps_request_sem, K_FOREVER);
+
+	err = modem_jwt_generate(&jwt);
+	if (err) {
+		LOG_ERR("Failed to generate JWT, err %d", err);
+		goto cleanup;
+	}
+	LOG_DBG("JWT:\n%s", log_strdup(jwt.jwt_buf));
+	rest_ctx.auth = jwt.jwt_buf;
+
+	LOG_INF("\n********************* A-GPS API *********************");
```

Comment:
you could remove this comment... it was useful for me when i was testing all the REST APIs in one sample.